fdbserver/MasterProxyServer.actor.cpp

```diff
@@ -1276,20 +1277,16 @@ ACTOR Future<Void> commitBatch(
 		ASSERT(p.second.isReady());
 	}
 
+	// After logging finishes, we report the commit version to master so that every other proxy can get the most
+	// up-to-date live committed version.
 	TEST(self->committedVersion.get() > commitVersion);   // A later version was reported committed first
+	if (SERVER_KNOBS->ASK_READ_VERSION_FROM_MASTER) {
+		wait(self->master.reportLiveCommittedVersion.getReply(ReportRawCommittedVersionRequest(commitVersion, lockedAfter, metadataVersionAfter), TaskPriority::ProxyMasterVersionReply));
```

Comment:
if commitVersion <= self->committedVersion.get(), do we **have to**  report the version? If we don't have to, this introduce an extra delay. I feel that the old code might work and you only need to remove the extra `if (commitVersion > self->committedVersion.get()) `

Reply:
After I did some checking, the old work did work which fixes errors related to watches and sideband. The extra `if (commitVersion > self->committedVersion.get()` is necessary because the committedVersion may become a larger value because the wait statement will yield the event loop and other commitBatch events may override the committedVersion to a large value.

This fix is to get rid of some wield errors such as slow/ConfigureTest.txt and slow/DDBalanceAndRemove.txt which are two Joshua errors after my previous fix ([PR_3428](~/fdb/foundationdb/tests/slow/DDBalanceAndRemove.txt)): 20200629-194014-apple::PR_3428-6ac65102cece631f. Joshua run on this fix don't have these two errors. (20200702-170456-apple::PR_3441-78bd28d028f9ccb5)

I used to suspect these two errors is related to my feature because these two targets don't fail before my change and fail after the change. Now I think they are probably not related to my feature because the seed is not stable between two commits whose diff has extra RPC involved.

Anyway, I will make this PR only enable this feature again.

Reply:
I agree configure and DD tests should not relate to your change.

Have you run 100K watches and sideband tests, just to make sure these two tests won't have failure when the feature is enabled? 
As discussed offline yesterday, if the failures we saw were due to the "race condition",  and the chance of the condition is low, then we might miss it if we only run 100K on all tests: When we simply kick off a Joshua, the tests on other unrelated tests (say the backup and restore) will also be counted into the 100K tests.

Reply:
The race condition will happen if the code block looks like this which is what the old code version used to be:
`      if( commitVersion > self->committedVersion.get() ) {
                self->committedVersion.set(commitVersion);
	        self->locked = lockedAfter;
		self->metadataVersion = metadataVersionAfter;
		if (SERVER_KNOBS->ASK_READ_VERSION_FROM_MASTER) {
			wait(self->master.reportLiveCommittedVersion.getReply(ReportRawCommittedVersionRequest(commitVersion, lockedAfter, metadataVersionAfter), TaskPriority::ProxyMasterVersionReply));
		}
	}`

This is how the race happens: suppose proxy1 is about to report two transactions (commit versions are v1 and v2, v1 > v2). Firstly, self->committedVersion is updated to v1 and waits for it to be reported to master. Then, the second txn comes and finds out that v2 is less than self->committedVersion, skip the report and claim the txn to be committed to the client. The client knows the txn is committed then immediately talks to proxy2 who then talks to master which may haven't received v1 and have a smaller value v3 (v3 < v2). Then the client gets a stale value v3 comparing to the committed version v2. 

Now we fixes it by reporting the commit version first and update self->committedVersion afterwards so that we can make sure  master's version >= self->committedVersion any time.